### PR TITLE
[Merged by Bors] - Reconstruct Payloads using Payload Bodies Methods

### PR DESCRIPTION
**Pull request by ethDreamer**
## Issue Addressed

* #3895

### CODE REVIEW

File: beacon_node/beacon_chain/src/beacon_block_streamer.rs [hunk 1]
```diff
@@ -0,0 +1,719 @@
...
+
+        // Now deal with the by_range requests. Sort them in order of increasing slot
+        let mut by_range = EngineRequest::<T::EthSpec>::new_by_range();
+        by_range_blocks.sort_by(|block_parts_a, block_parts_b| {
+            // this unwrap shouldn't occur as slot is never u64::NAN
+            block_parts_a
+                .slot()
+                .partial_cmp(&block_parts_b.slot())
+                .unwrap_or(std::cmp::Ordering::Equal)
+        });
```
Comment (michaelsproul) [inline beacon_node/beacon_chain/src/beacon_block_streamer.rs]: `Slot` implements `Ord` so we can use `sort_by_key`:

```suggestion
        by_range_blocks.sort_by_key(|block_parts| block_parts.slot());
```

**Review by michaelsproul — COMMENTED** [review]
Looking good overall. I'm still in the process of understanding the block streamer and will come back for a more thorough review next week.

Some thoughts on complexity:

- The current impl handles a graceful conversion from BlocksByRoot requests to PayloadsByRange. I initially thought this might not be necessary and we could always convert BlocksByRoot to PayloadsByRoot, but the non-finality requirement for using PayloadsByRoot makes that impossible.
- I'd suggested we could avoid ever using PayloadsByRoot because all unfinalized payloads _should_ be present in the hot database. This works OK in the case where the EL's view of finality is up-to-date or ahead of the CL, but _not_ in the case where the EL's view is behind. In practice I think handling the case where the EL is behind might be important, because the CL can sync ahead using optimistic sync. In this case however the CL's view of the finalized slot is irrelevant to whether or not we _should_ use PayloadsByRoot, so perhaps we should refactor to _always use_ PayloadsByRange and treat PayloadsByRoot as a fallback (like getBlockByHash).

**Comment by ethDreamer** [comment]
I've added some more descriptive logging to the streamer (perhaps we should add metrics?). Looks like most requests are for ~64 blocks. I monitored the logs for ~5 hours. In that time, every single block requested was returned. The overwhelming majority of requests require either `0`,`1`, or `2` requests to the EE. There were three notable exceptions to this however, not sure what causes them:

```
Mar 08 20:09:51.953 DEBG BeaconBlockStreamer finished            engine requests: 16, failed: 0, succeeded: 510, sent: 510, requested blocks: 510, service: beacon
...
Mar 08 20:49:15.203 DEBG BeaconBlockStreamer finished            engine requests: 4, failed: 0, succeeded: 127, sent: 127, requested blocks: 127, service: beacon
...
Mar 08 22:51:07.699 DEBG BeaconBlockStreamer finished            engine requests: 16, failed: 0, succeeded: 509, sent: 509, requested blocks: 509, service: beacon
```
here's a random section of continuous logs over the course of ~5 minutes:
```
Mar 08 20:30:00.318 DEBG BeaconBlockStreamer finished            engine requests: 2, failed: 0, succeeded: 63, sent: 63, requested blocks: 63, service: beacon
Mar 08 20:30:03.579 DEBG BeaconBlockStreamer finished            engine requests: 2, failed: 0, succeeded: 64, sent: 64, requested blocks: 64, service: beacon
Mar 08 20:30:21.724 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 64, sent: 64, requested blocks: 64, service: beacon
Mar 08 20:30:28.218 DEBG BeaconBlockStreamer finished            engine requests: 2, failed: 0, succeeded: 62, sent: 62, requested blocks: 62, service: beacon
Mar 08 20:30:28.241 DEBG BeaconBlockStreamer finished            engine requests: 2, failed: 0, succeeded: 64, sent: 64, requested blocks: 64, service: beacon
Mar 08 20:30:49.255 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 1, sent: 1, requested blocks: 1, service: beacon
Mar 08 20:30:53.154 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 64, sent: 64, requested blocks: 64, service: beacon
Mar 08 20:31:02.356 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 64, sent: 64, requested blocks: 64, service: beacon
Mar 08 20:31:35.398 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 63, sent: 63, requested blocks: 63, service: beacon
Mar 08 20:31:37.064 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 64, sent: 64, requested blocks: 64, service: beacon
Mar 08 20:31:38.663 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 64, sent: 64, requested blocks: 64, service: beacon
Mar 08 20:31:41.035 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 63, sent: 63, requested blocks: 63, service: beacon
Mar 08 20:31:42.317 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 64, sent: 64, requested blocks: 64, service: beacon
Mar 08 20:31:46.294 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 64, sent: 64, requested blocks: 64, service: beacon
Mar 08 20:31:47.889 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 1, sent: 1, requested blocks: 1, service: beacon
Mar 08 20:31:48.154 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 1, sent: 1, requested blocks: 1, service: beacon
Mar 08 20:31:48.310 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 1, sent: 1, requested blocks: 1, service: beacon
Mar 08 20:31:48.438 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 1, sent: 1, requested blocks: 1, service: beacon
Mar 08 20:31:48.558 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 1, sent: 1, requested blocks: 1, service: beacon
Mar 08 20:31:48.709 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 1, sent: 1, requested blocks: 1, service: beacon
Mar 08 20:31:48.853 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 1, sent: 1, requested blocks: 1, service: beacon
Mar 08 20:31:49.066 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 1, sent: 1, requested blocks: 1, service: beacon
Mar 08 20:31:49.346 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 1, sent: 1, requested blocks: 1, service: beacon
Mar 08 20:31:49.508 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 1, sent: 1, requested blocks: 1, service: beacon
Mar 08 20:31:49.653 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 1, sent: 1, requested blocks: 1, service: beacon
Mar 08 20:31:49.803 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 1, sent: 1, requested blocks: 1, service: beacon
Mar 08 20:31:50.068 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 1, sent: 1, requested blocks: 1, service: beacon
Mar 08 20:31:50.280 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 1, sent: 1, requested blocks: 1, service: beacon
Mar 08 20:31:50.416 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 1, sent: 1, requested blocks: 1, service: beacon
Mar 08 20:31:50.612 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 1, sent: 1, requested blocks: 1, service: beacon
Mar 08 20:31:50.774 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 1, sent: 1, requested blocks: 1, service: beacon
Mar 08 20:31:50.926 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 1, sent: 1, requested blocks: 1, service: beacon
Mar 08 20:31:51.082 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 1, sent: 1, requested blocks: 1, service: beacon
Mar 08 20:31:51.197 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 1, sent: 1, requested blocks: 1, service: beacon
Mar 08 20:31:51.205 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 64, sent: 64, requested blocks: 64, service: beacon
Mar 08 20:31:51.346 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 1, sent: 1, requested blocks: 1, service: beacon
Mar 08 20:31:51.490 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 1, sent: 1, requested blocks: 1, service: beacon
Mar 08 20:31:51.666 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 1, sent: 1, requested blocks: 1, service: beacon
Mar 08 20:31:51.822 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 1, sent: 1, requested blocks: 1, service: beacon
Mar 08 20:31:51.973 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 1, sent: 1, requested blocks: 1, service: beacon
Mar 08 20:31:52.238 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 1, sent: 1, requested blocks: 1, service: beacon
Mar 08 20:31:52.422 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 1, sent: 1, requested blocks: 1, service: beacon
Mar 08 20:31:55.508 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 64, sent: 64, requested blocks: 64, service: beacon
Mar 08 20:31:59.240 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 64, sent: 64, requested blocks: 64, service: beacon
Mar 08 20:32:01.710 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 64, sent: 64, requested blocks: 64, service: beacon
Mar 08 20:32:03.346 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 63, sent: 63, requested blocks: 63, service: beacon
Mar 08 20:32:06.802 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 60, sent: 60, requested blocks: 60, service: beacon
Mar 08 20:32:13.114 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 64, sent: 64, requested blocks: 64, service: beacon
Mar 08 20:32:18.552 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 64, sent: 64, requested blocks: 64, service: beacon
Mar 08 20:32:24.376 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 63, sent: 63, requested blocks: 63, service: beacon
Mar 08 20:32:26.488 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 61, sent: 61, requested blocks: 61, service: beacon
Mar 08 20:32:28.016 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 64, sent: 64, requested blocks: 64, service: beacon
Mar 08 20:32:31.121 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 64, sent: 64, requested blocks: 64, service: beacon
Mar 08 20:32:46.479 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 64, sent: 64, requested blocks: 64, service: beacon
Mar 08 20:32:48.334 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 64, sent: 64, requested blocks: 64, service: beacon
Mar 08 20:32:53.455 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 63, sent: 63, requested blocks: 63, service: beacon
Mar 08 20:32:54.810 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 63, sent: 63, requested blocks: 63, service: beacon
Mar 08 20:33:05.692 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 64, sent: 64, requested blocks: 64, service: beacon
Mar 08 20:33:13.121 DEBG BeaconBlockStreamer finished            engine requests: 2, failed: 0, succeeded: 64, sent: 64, requested blocks: 64, service: beacon
Mar 08 20:33:17.148 DEBG BeaconBlockStreamer finished            engine requests: 2, failed: 0, succeeded: 63, sent: 63, requested blocks: 63, service: beacon
Mar 08 20:33:32.995 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 64, sent: 64, requested blocks: 64, service: beacon
Mar 08 20:33:37.530 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 64, sent: 64, requested blocks: 64, service: beacon
Mar 08 20:33:55.265 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 1, sent: 1, requested blocks: 1, service: beacon
Mar 08 20:34:37.183 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 1, sent: 1, requested blocks: 1, service: beacon
Mar 08 20:34:49.910 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 64, sent: 64, requested blocks: 64, service: beacon
Mar 08 20:34:58.801 DEBG BeaconBlockStreamer finished            engine requests: 2, failed: 0, succeeded: 62, sent: 62, requested blocks: 62, service: beacon
Mar 08 20:35:05.640 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 64, sent: 64, requested blocks: 64, service: beacon
Mar 08 20:35:14.421 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 64, sent: 64, requested blocks: 64, service: beacon
Mar 08 20:35:25.239 DEBG BeaconBlockStreamer finished            engine requests: 1, failed: 0, succeeded: 32, sent: 32, requested blocks: 32, service: beacon
Mar 08 20:35:28.087 DEBG BeaconBlockStreamer finished            engine requests: 2, failed: 0, succeeded: 63, sent: 63, requested blocks: 63, service: beacon
Mar 08 20:35:31.357 DEBG BeaconBlockStreamer finished            engine requests: 0, failed: 0, succeeded: 64, sent: 64, requested blocks: 64, service: beacon
```

**Comment by michaelsproul** [comment]
In my previous review I wrote:

> the CL's view of the finalized slot is irrelevant to whether or not we should use PayloadsByRoot, so perhaps we should refactor to always use PayloadsByRange and treat PayloadsByRoot as a fallback (like getBlockByHash).

Having thought about it more I'm quite certain we should action this. The current approach of trying to use `ByHash` if the block's slot is after the finalized slot is faulty because we never actually expect that check to pass: if a block is stored in Lighthouse's database _without a payload_ then it _must_ be prior to the database's split slot, which is <= the finalized slot from fork choice, i.e. the `else` branch here is unreachable:

https://github.com/sigp/lighthouse/blob/9a7d32fe524a78ee1af644bc1e5b5bb1824ee570/beacon_node/beacon_chain/src/beacon_block_streamer.rs#L643-L652

The fundamental reason for this is that we always store full payloads for all blocks in the hot part of the database (with `slot >= split`).

I think there are two approaches we could take to remedy this:

1) Not using `ByHash` at all. Try `ByRange` and if it fails, let it fail. If Lighthouse's view of finality is ahead of the execution node then arguably we are still in a syncing state and shouldn't be burdening the EL with payload reconstruction requests. If all is well the EL will catch up soon enough and we'll be able to return payloads.
2) Use `ByHash` only after observing a failure of `ByRange`. This is trickier because we need to hold on to the `BlockParts` from the `ByRange` request to re-use them in the fallback `ByHash` request. We would also need (?) to do some batching of the `ByHash` requests to avoid spamming too many individual requests. A simple algorithm would be to batch consecutive errors into a single `ByHash` request.

I started trying to implement (2) with fallback logic [at this point](https://github.com/sigp/lighthouse/blob/9a7d32fe524a78ee1af644bc1e5b5bb1824ee570/beacon_node/beacon_chain/src/beacon_block_streamer.rs#L780) but switched to (1) when I encountered the `BlockPart`s issue. The two commits I made are:

- General cleanup commit/refactor (relevant to both approaches): 35d6ab82a31402e3e2c624be09c6f6f2ae611a0d.
- Deletion of `ByHash`: 2190ef10680734fcca539d77f1cffe3410017f61.

One option would be to take approach (1) for now, and come back for (2) at a later date.

**Comment by ethDreamer** [comment]
Looks good to me! I've merged your changes in :)

**Review by michaelsproul — APPROVED** [review]
I'm happy to merge this to `unstable`!

I've added some metrics and this PR is running on both Sepolia + Goerli without issue :ok_hand:

**Comment by michaelsproul** [comment]
I've seen no issues running this on our testnet infra, merging! :tada:

**Pull request closed without merging**